x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java

```diff
@@ -1266,6 +1284,241 @@ private boolean foldable(Expression e) {
+                        try {
+                            // try date math first
+                            Tuple<Long, Long> range = DateFieldMapper.DateFieldType.toTimestampRange(
+                                foldedValue,
+                                foldedValue,
+                                true,
+                                true,
+                                ZoneOffset.UTC,
+                                DateUtils.DATE_MATH_PARSER,
+                                DateFieldMapper.Resolution.MILLISECONDS,
+                                () -> System.currentTimeMillis()
+                            );
+                            if (range.v1() < range.v2()) {
+                                /*
+                                treat it as a date range
+                                eg. the following
+                                something IN (<daterange>, <somethingElse1>, <somethingElse2>)
+                                becomes
+                                something BETWEEN <min(daterange)> AND <max(daterange)> OR someting IN (<somethingElse1>, <somethingElse2>)
+                                */
+                                additionalConditions.add(
+                                    new Range(
+                                        in.source(),
+                                        in.value(),
+                                        tryCast(exp, range.v1(), type),
+                                        true,
+                                        tryCast(exp, range.v2(), type),
+                                        true,
+                                        in.zoneId()
+                                    )
+                                );
+                            } else {
+                                newRight.add(tryCast(exp, foldedValue, DataTypes.DATETIME));
+                            }
+                        } catch (ElasticsearchParseException | DateTimeParseException e) {
+                            // the date parsing just failed, use the original expression
+                            newRight.add(exp);
+                        }
+                    } else {
+                        newRight.add(tryCast(exp, foldedValue, type));
+                    }
+                }
+            }
+            if (converted == false) {
+                return in;
+            }
+            if (newRight.size() == 1) {
+                additionalConditions.add(new Equals(in.source(), in.value(), newRight.remove(0)));
+            }
+
+            if (newRight.size() > 0 && additionalConditions.size() > 0) {
+                return new Or(
+                    in.source(),
+                    new In(in.source(), in.value(), newRight, in.zoneId()),
+                    toOrTree(in.source(), additionalConditions)
+                );
+            } else if (newRight.size() > 0) {
+                return new In(in.source(), in.value(), newRight, in.zoneId());
```

Comment:
Could the IN have been modified here? I.e. shouldn't the original `in` be returned?

Reply:
A few lines before there is an explicit check:

```
            if (converted == false) {
                return in;
            }
```

If the method reached this point, it means there were some conversions (probably not date math conversions, otherwise there would be some `additionalConditions`), so the `newRight` list of values should be different from the original one.